platform/engineering/design-docs/2021-01-13-web-component-library.md

```diff
@@ -0,0 +1,221 @@
+
+For now each Component will have its own styles _implicitly_ based off of USWDS, which resemble what is currently available in the Design System. Potential future work would involve explicitly importing USWDS component stylesheets into each Component's CSS file.
+
+#### Loading Web Components on va.gov
+
+The setup code for the WC library will be in its own `vets-website` entry file that is loaded on each page. This code will be responsible for [applying polyfills](https://stenciljs.com/docs/faq#what-polyfills-does-stencil-provide) for older browsers as well as [defining the Web Components](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define).
+
+In addition, this script will be one of the first loaded in order to ensure that the Components will be defined in the browser before they are rendered on the page. This will be done by adding a `<script>` entry towards the top of [the `<head>` tag that gets loaded on every page](https://github.com/department-of-veterans-affairs/vets-website/blob/dc5018065eccf073d172097a19922755126c98a3/src/site/includes/header.html#L16)
+
+For further optimization, the Components will be lazy-loaded. Even if our Web Component library ends up having hundreds of Components, lazy-loading will will reduce the risk of a significant negative performance impact. Here is a summary of the process:
+
+1. each Component is initially defined as a wrapper
+1. one of many Components appears in the DOM
+1. wrapper Component lazy-loads the _actual_ definition
+1. properly defined Component is rendered in the browser
+
+[Stencil has a useful article that talks about their lazy loading approach](https://stenciljs.com/blog/how-lazy-loading-web-components-work).
+
+> It doesn’t matter how the element came to be rendered in the DOM, either way, Stencil knows how to lazy-load itself using web-standards and native browser APIs. And by deferring most of the logic to the browser, rather than generating more JavaScript runtime, the components themselves have an optimal file size.
+
+#### Migration to Web Components
+
+Once a new release is made for the Web Component library, we will [configure a linting rule](https://github.com/AlexMost/eslint-plugin-deprecate/blob/master/docs/rules/import.md) as part of the `lint:js:changed` script in `vets-website` to throw an error if a React component is imported that has a released Web Component replacement. This will prevent the older React component from appearing in more places as we are actively making efforts to remove it.
```

Comment:
:thinking: _Can_ we add this linting rule only for the `lint:js:changed` NPM script?

With this process, if an engineer makes _any_ changes to a file that imports the legacy component, they'll be forced to update the component usage before committing (or ignore the commit hook linting), right?

For the most part, forcing an update is probably fine if we're not changing the API of the component. If it's just like a `<TextInput>` -> `<va-text-input>` rename, great, but if it'll require more involved changes, that's probably not awesome.

How might we reduce the burden for engineers making unrelated changes? Is there any way to catch only _new_ imports for legacy components?

Reply:
> _Can_ we add this linting rule only for the lint:js:changed NPM script?

Yes, but that will likely involve some minor refactoring of the current config and/or the script itself. One option would be to change [the current command](https://github.com/department-of-veterans-affairs/vets-website/blob/9da6a50293000210257703ffef135c10fb6fe6d9/package.json#L38) from `eslint --quiet` to something like:

```
eslint --quiet -c component-library.eslint.js
```

Where `component-library.eslint.js` is [a custom configuration file](https://eslint.org/docs/user-guide/configuring#configuration-file-formats) that imports the existing `.eslintrc` and exports it as json with our new rule added along with the export.

---

> With this process, if an engineer makes any changes to a file that imports the legacy component, they'll be forced to update the component usage before committing (or ignore the commit hook linting), right?

I believe so - that is the limitation of our current pre-commit linting. It lints entire changed files, not just the changed chunks.

---

> How might we reduce the burden for engineers making unrelated changes? Is there any way to catch only new imports for legacy components?

I mean... yeah, it's a solvable problem. We could capture the output from the `eslint` command, and if there is an error we could:

1. See if it is related to a `component-library` import
1. Check the git log or blame and see if the commit is before or after `HEAD`

Do you think this is something we need to put the time into?

Reply:
Cool. Let's work with the FE tools team on this when the time comes. We can iterate on the linting even after this doc gets approved.

cc @Clark-W-Griswold @meganhkelley for visibility